application/forms/EntryForm.php

```diff
@@ -311,114 +317,233 @@ protected function assemble()
+            ->filter(
+                Filter::all(
+                    Filter::equal('timeperiod.owned_by_schedule_id', $scheduleId),
+                    Filter::equal('schedule_member.membership_hash', $calculatedHash)
+                )
+            );
 
-        $db->update('timeperiod_entry', $data, ['id = ?' => $id]);
+        $prevTimeperiodEntriesCount = TimeperiodEntry::on($db)
+            ->with('timeperiod_entry.timeperiod')
+            ->filter(Filter::equal('timeperiod.owned_by_schedule_id', $scheduleId))
+            ->filter(Filter::equal('timeperiod_id', $prevTimeperiodId))
+            ->count();
 
-        $members = ScheduleMember::on($db)
-            ->filter(Filter::all(
-                Filter::equal('schedule_id', $scheduleId),
-                Filter::equal('timeperiod_id', $timeperiodId)
-            ));
+        $db->beginTransaction();
 
-        $recipients = explode(',', $this->getValue('recipient'));
+        if ($changedMembers->count() > 0) {
+            // Update the entries with timeperiod_id of the changed membership
+            $db->update(
+                'timeperiod_entry',
+                $data + ['timeperiod_id' => (int) $changedMembers->first()->timeperiod_id],
+                ['id = ?' => $id]
+            );
+            $prevTimeperiodEntriesCount -= 1;
+        } elseif ($prevTimeperiodEntriesCount === 1) {
+            // Update the membership hash and add or remove members for the combination of existing
+            // schedule_id and timeperiod_id
+            $recipients = explode(',', $this->getValue('recipient'));
 
-        $users = [];
-        $groups = [];
-        foreach ($recipients as $recipient) {
-            list($type, $id) = explode(':', $recipient, 2);
+            $users = [];
+            $groups = [];
+            foreach ($recipients as $recipient) {
+                list($type, $id) = explode(':', $recipient, 2);
 
-            if ($type === 'contact') {
-                $users[$id] = $id;
-            } elseif ($type === 'group') {
-                $groups[$id] = $id;
+                if ($type === 'contact') {
+                    $users[$id] = $id;
+                } elseif ($type === 'group') {
+                    $groups[$id] = $id;
+                }
             }
-        }
 
-        $usersToRemove = [];
-        $groupsToRemove = [];
-        foreach ($members as $member) {
-            if ($member->contact_id !== null) {
-                if (! isset($users[$member->contact_id])) {
-                    $usersToRemove[] = $member->contact_id;
-                } else {
-                    unset($users[$member->contact_id]);
-                }
-            } else {
-                if (! isset($groups[$member->contactgroup_id])) {
-                    $groupsToRemove[] = $member->contactgroup_id;
+            $usersToRemove = [];
+            $groupsToRemove = [];
+            $prevMembers = ScheduleMember::on(Database::get())
```

Comment:
With this, you'll fetch members of other timeperiods that happen to match the hash by accident.

Reply:
Oh, never mind...